### PR TITLE
Exception handling

### PR DESCRIPTION
**Pull request by axthosarouris**
Changes:
1. Log causes,
2. Log stacktrace
3. All the log in one line

### CODE REVIEW

File: src/main/java/nva/commons/hanlders/ApiGatewayHandler.java [hunk 1]
```diff
@@ -208,24 +216,59 @@ protected void writeOutput(I input, O output)
...
+     */
+    protected void writeFailure(Exception exception, Integer statusCode, String additionalMessage) throws IOException {
         try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(outputStream))) {
             String outputString = Optional.ofNullable(exception.getMessage()).orElse(defaultErrorMessage());
+            outputString = addAdditionalMessage(additionalMessage, outputString);
             GatewayResponse<String> gatewayResponse =
-                new GatewayResponse<>(outputString, getFailureHeaders(), getFailureStatusCode(input, exception));
+                new GatewayResponse<>(outputString, getFailureHeaders(), statusCode);
             String gateWayResponseJson = objectMapper.writeValueAsString(gatewayResponse);
             writer.write(gateWayResponseJson);
```
Comment (janbj) [inline src/main/java/nva/commons/hanlders/ApiGatewayHandler.java:248]: Can you use objectMapper.write(outputStream, gatewayResponse) directly?

File: src/main/java/nva/commons/hanlders/ApiGatewayHandler.java [hunk 2]
```diff
@@ -208,24 +216,59 @@ protected void writeOutput(I input, O output)
...
+     * @param additionalMessage any additional message that is necessary to send. Set to null for no additional message
+     * @throws IOException when the writer throws an IOException.
+     */
+    protected void writeFailure(Exception exception, Integer statusCode, String additionalMessage) throws IOException {
         try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(outputStream))) {
             String outputString = Optional.ofNullable(exception.getMessage()).orElse(defaultErrorMessage());
+            outputString = addAdditionalMessage(additionalMessage, outputString);
             GatewayResponse<String> gatewayResponse =
-                new GatewayResponse<>(outputString, getFailureHeaders(), getFailureStatusCode(input, exception));
+                new GatewayResponse<>(outputString, getFailureHeaders(), statusCode);
```
Comment (janbj) [inline src/main/java/nva/commons/hanlders/ApiGatewayHandler.java:246]: Is this formatted as a Problem JSON?

File: src/main/java/nva/commons/hanlders/ApiGatewayHandler.java [hunk 3]
```diff
@@ -245,14 +288,45 @@ public void handleRequest(InputStream input, OutputStream output, Context contex
             writeOutput(inputObject, response);
         } catch (ApiGatewayException e) {
             logger.log(e.getMessage());
```
Comment (janbj) [inline src/main/java/nva/commons/hanlders/ApiGatewayHandler.java:290]: Do we log the RequestId from the AWS Context? We should probably include it in every log line and return it in the Problem in the response.

File: src/test/java/nva/commons/handlers/ApiGatewayHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    @BeforeEach
+    public void setup() {
+        environment = mock(Environment.class);
+        when(environment.readEnv(anyString())).thenReturn(SOME_ENV_VALUE);
+        context = mock(Context.class);
+        when(context.getLogger()).thenReturn(logger);
+    }
+
+    @Test
+    public void handleRequestShouldHaveAvailableTheRequestHeaders() throws IOException {
```
Comment (janbj) [inline src/test/java/nva/commons/handlers/ApiGatewayHandlerTest.java:64]: DisplayName?

File: src/test/java/nva/commons/handlers/ApiGatewayHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        assertThat(logs, containsString(IllegalStateException.class.getName()));
+        assertThat(logs, containsString(IllegalArgumentException.class.getName()));
+        assertThat(logs, containsString(IllegalCallerException.class.getName()));
+        assertThat(logs, containsString(FIRST_EXCEPTION_MESSAGE));
+    }
+
+    private Handler handlerThatThrowsUncheckedExceptions() {
+        return new Handler(environment) {
+            @Override
+            protected String processInput(RequestBody input, RequestInfo requestInfo, Context context)
```
Comment (janbj) [inline src/test/java/nva/commons/handlers/ApiGatewayHandlerTest.java:140]: Mockito.when().thenThrow() ?

File: src/main/java/nva/commons/hanlders/ApiGatewayHandler.java (hunk 2 above)
Comment (axthosarouris) [reply src/main/java/nva/commons/hanlders/ApiGatewayHandler.java:246]: No this is the next PR

File: src/main/java/nva/commons/hanlders/ApiGatewayHandler.java (hunk 1 above)
Comment (axthosarouris) [reply src/main/java/nva/commons/hanlders/ApiGatewayHandler.java:248]: I replace this code in the next PR. This PR is focused more on the changes  on the API. The output of the failure (Zalando problem) is handled on antoher PR

File: src/test/java/nva/commons/handlers/ApiGatewayHandlerTest.java (hunk 5 above)
Comment (axthosarouris) [reply src/test/java/nva/commons/handlers/ApiGatewayHandlerTest.java:142]: I need to implement all the other methods. (handleRequest, writeFailure, etc)

File: src/main/java/nva/commons/hanlders/ApiGatewayHandler.java (hunk 3 above)
Comment (axthosarouris) [reply src/main/java/nva/commons/hanlders/ApiGatewayHandler.java:290]: I agree. But the code of this method is replaced in the next PR. I have changed it there

**Pull request merged**